pkg/gadgets/profile/tracer/core/tracer.go

```diff
@@ -0,0 +1,388 @@
+// limitations under the License.
+
+package tracer
+
+// #include <linux/types.h>
+// #include "./bpf/profile.h"
+import "C"
+
+import (
+	"bufio"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"runtime"
+	"sort"
+	"strconv"
+	"strings"
+	"unsafe"
+
+	log "github.com/sirupsen/logrus"
+
+	"github.com/cilium/ebpf"
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgets"
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgets/profile/tracer"
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgets/profile/types"
+	"golang.org/x/sys/unix"
+)
+
+//go:generate sh -c "GOOS=$(go env GOHOSTOS) GOARCH=$(go env GOHOSTARCH) go run github.com/cilium/ebpf/cmd/bpf2go -target bpfel -cc clang profile ./bpf/profile.bpf.c -- -I./bpf/ -I../../../../ -target bpf -D__TARGET_ARCH_x86"
+
+type Tracer struct {
+	resolver gadgets.Resolver
+	node     string
+	objs     profileObjects
+	perfFds  []int
+	config   *tracer.Config
+}
+
+const (
+	perfMaxStackDepth = 127
+	perfSampleFreq    = 49
+	// In C, struct perf_event_attr has a freq field which is a bit in a
+	// 64-length bitfield.
+	// In Golang, there is a Bits field which 64 bits long.
+	// From C, we can deduce freq (which permits using frequency not period)
+	// is the 10th bit.
+	frequencyBit = 1 << 10
+)
+
+func NewTracer(resolver gadgets.Resolver, config *tracer.Config, node string) (*Tracer, error) {
+	t := &Tracer{
+		resolver: resolver,
+		node:     node,
+		config:   config,
+	}
+
+	if err := t.start(); err != nil {
+		t.Stop()
+		return nil, err
+	}
+
+	return t, nil
+}
+
+type keyCount struct {
+	key   C.struct_key_t
+	value uint64
+}
+
+func readCountsMap(counts *ebpf.Map) ([]keyCount, error) {
+	var prev *C.struct_key_t = nil
+	keysCounts := []keyCount{}
+	key := C.struct_key_t{}
+
+	i := 0
+	for {
+		if err := counts.NextKey(unsafe.Pointer(prev), unsafe.Pointer(&key)); err != nil {
+			if errors.Is(err, ebpf.ErrKeyNotExist) {
+				break
+			}
+			return nil, fmt.Errorf("error getting next key: %w", err)
+		}
+
+		var value uint64
+		err := counts.Lookup(key, unsafe.Pointer(&value))
+		if err != nil {
+			return nil, err
+		}
+
+		kv := keyCount{
+			key:   key,
+			value: value,
+		}
+
+		if i < len(keysCounts)-1 {
+			keysCounts[i] = kv
+		} else {
+			keysCounts = append(keysCounts, kv)
+		}
+
+		if value == 0 {
+			continue
+		}
+
+		prev = &key
+		i++
+	}
+
+	return keysCounts, nil
+}
+
+type kernelSymbol struct {
+	addr uint64
+	name string
+}
+
+// readKernelSymbols reads /proc/kallsyms and a slice of kernelSymbols.
+func readKernelSymbols() ([]kernelSymbol, error) {
+	symbols := []kernelSymbol{}
+
+	file, err := os.Open("/proc/kallsyms")
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		line := scanner.Text()
+
+		fields := strings.Fields(line)
+		if len(fields) < 3 {
+			continue
+		}
+
+		addr, err := strconv.ParseUint(fields[0], 16, 64)
+		if err != nil {
+			return nil, err
```

Comment:
nit: Is it intended that this returns an error here but in L149 it skips the current line?